src/backend/cuda/kernel/memcopy.hpp

```diff
@@ -17,86 +17,248 @@
+                }
+                mndims = AF_MAX_DIMS;
+                break;  // Perform this operation only once.
+            }
+        }
+    }
+}
 
-    if (ndims == 1) {
-        threads.x *= threads.y;
-        threads.y = 1;
+template<typename T>
+void memcopy(Param<T> out, CParam<T> in, dim_t indims) {
+    bool isLinear  = true;
+    dim_t elements = (indims == 0) ? 0 : 1;
+    for (dim_t dim = 0; dim < indims; ++dim) {
+        isLinear &=
+            (elements == in.strides[dim]) & (elements == out.strides[dim]);
+        elements *= in.dims[dim];
     }
+    if (elements > 0) {
+        const auto stream = cuda::getActiveStream();
+        if (isLinear) {
+            CUDA_CHECK(cudaMemcpyAsync(out.ptr, in.ptr,
+                                       in.elements() * sizeof(T),
+                                       cudaMemcpyDeviceToDevice, stream));
+        } else {
+            const int *maxGridSize =
+                cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize;
+            serializeArray(in.dims, in.strides, indims, out.strides);
+            increaseWorkload(elements, in.dims, in.strides, indims,
+                             out.strides);
+            const dim3 threads = bestBlockSize<dim3>(in.dims, 32);
+            dim3 blocks(divup(static_cast<unsigned>(in.dims[0]), threads.x),
+                        divup(static_cast<unsigned>(in.dims[1]), threads.y),
+                        divup(static_cast<unsigned>(in.dims[2]), threads.z));
+            const bool loop1 = blocks.y > static_cast<unsigned>(maxGridSize[1]);
+            if (loop1) blocks.y = maxGridSize[1];
+            const bool loop2 = blocks.z > static_cast<unsigned>(maxGridSize[2]);
+            if (loop2) blocks.z = maxGridSize[2];
+            EnqueueArgs qArgs(blocks, threads, stream);
 
-    // FIXME: DO more work per block
-    uint blocks_x = divup(in.dims[0], threads.x);
-    uint blocks_y = divup(in.dims[1], threads.y);
+            auto memCopy = common::getKernel("cuda::memcopy", {memcopy_cuh_src},
+                                             {
+                                                 TemplateTypename<T>(),
+                                                 TemplateArg(loop1),
+                                                 TemplateArg(loop2),
+                                             });
+            memCopy(qArgs, out, in);
+            POST_LAUNCH_CHECK();
+        }
+    }
+}
 
-    dim3 blocks(blocks_x * in.dims[2], blocks_y * in.dims[3]);
+template<typename T>
+class CParamPlus {
+   public:
+    CParam<T> cparam;
+    dim_t ooffset;
+    CParamPlus(CParam<T> arr, dim_t off) : cparam(arr), ooffset(off){};
+};
 
-    const int maxBlocksY =
-        cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize[1];
-    blocks.z = divup(blocks.y, maxBlocksY);
-    blocks.y = divup(blocks.y, blocks.z);
+template<typename T>
+void memcopyN(Param<T> out, std::vector<CParamPlus<T>> &Ins) {
```

Comment:
Calling the memcopy for each array separately means that the overhead has to be processed for each copy individually.
This was the model I had in the beginning.

After moving to memcopyN, I noticed a measurable difference (especially for small array's, which is typical for addBias() functions).

In case you give priority to this complexity reduction, **please let me know**.
I know alternative for addBias() performing similar to current join implementation.
```
af::array addBias(const af::array in) {
    af::dim4 dimsR(in.dims());
    ++dimsR[0];
    af:: array R = af::constant(1.0, dimsR);                    // yes full array, since very fast.
    R(af::seq(1,af::end),af::span,af::span,af::span) = in;   // overwrite for values of in
    return(R);
}
```